R/bundlePython.R

```diff
@@ -0,0 +1,134 @@
+    withCallingHandlers(
+      inferPythonEnv(
+        appDir,
+        python = python,
+        condaMode = condaMode,
+        forceGenerate = forceGenerate
+      ),
+      error = function(err) {
+        cli::cli_abort(
+          "Failed to detect python environment",
+          parent = err
+        )
+      }
+    )
+  }
+}
+
+appUsesPython <- function(quartoInfo) {
+  if (is.null(quartoInfo)) {
+    # No R-based, non-Quarto content uses Python by default.
+    # Looking for Python chunks in Rmd needs to happen separately.
+    FALSE
+  } else {
+    # Python is a direct consequence of the "jupyter" engine; not "knitr" or "markdown".
+    "jupyter" %in% quartoInfo[["engines"]]
+  }
+}
+
+getPythonForTarget <- function(path, accountDetails) {
+  # python is enabled on Connect and posit.cloud, but not on Shinyapps
+  targetIsShinyapps <- isShinyappsServer(accountDetails$server)
+  pythonEnabled <- getOption(
+    "rsconnect.python.enabled",
+    default = !targetIsShinyapps
+  )
+  if (pythonEnabled) {
+    getPython(path)
+  } else {
+    NULL
+  }
+}
+
+getPython <- function(path = NULL) {
+  if (is.null(path)) {
+    path <- Sys.getenv("RETICULATE_PYTHON", unset = Sys.getenv("RETICULATE_PYTHON_FALLBACK"))
+    if (path == "") {
+      return(NULL)
+    }
+  }
+  path.expand(path)
+}
+
+inferPythonEnv <- function(workdir,
+                           python = getPython(),
+                           condaMode = FALSE,
+                           forceGenerate = FALSE) {
+  # run the python introspection script
+  env_py <- system.file("resources/environment.py", package = "rsconnect")
+  args <- c(
+    shQuote(env_py),
+    if (condaMode) "-c",
+    if (forceGenerate) "-f",
+    shQuote(workdir)
+  )
+
+  # First check for reticulate. Then see if python is loaded in reticulate
+  # space, verify anaconda presence, and verify that the user hasn't specified
+  # that they don't want their conda environment captured.
+  hasConda <- is_installed("reticulate") &&
+    reticulate::py_available(initialize = FALSE) &&
+    reticulate::py_config()$anaconda &&
+    !condaMode
+
+  if (hasConda) {
+    prefix <- getCondaEnvPrefix(python)
+    conda <- getCondaExeForPrefix(prefix)
+    args <- c("run", "-p", prefix, python, args)
+    # conda run -p <prefix> python inst/resources/environment.py <flags> <dir>
+    output <- system2(command = conda, args = args, stdout = TRUE, stderr = NULL, wait = TRUE)
+  } else {
+    output <- system2(command = python, args = args, stdout = TRUE, stderr = NULL, wait = TRUE)
+  }
+
+  environment <- jsonlite::fromJSON(output)
+  if (is.null(environment$error)) {
+
+    list(
+      version = environment$python,
+      package_manager = list(
+        name = environment$package_manager,
+        version = environment[[environment$package_manager]],
+        package_file = environment$filename,
+        contents = environment$contents
+      )
+    )
+  } else {
+    cli::cli_abort(environment$error)
```

Comment:
Can you suggest a way to test this branch of code? i.e. is there a simple scenario that will cause `resources/environment.py` to fail?

Reply:
Unfortunately, no. @mmarchetti - could you help create a scenario to exercise the error conditions, here?

Reply:
Possible causes:
* An error running `python -m pip --version`
* An error running `python -m pip list --format=freeze`
* `requirements.txt` exists but is not readable

Probably the last one would be the easiest condition to create.

Reply:
@mmarchetti you mean like create it and then chmod 000 it? Or can I just create something that doesn't parse?

Reply:
Chmod 000 worked for me.